### PR TITLE
Exploratory: Enable volumes mounting in functions

### PR DESCRIPTION
**Pull request by martindekov**
Enabling volumes to be mounted in function's file
system using annotations

Signed-off-by: Martin Dekov <mdekov@vmware.com>

# Disclaimer

*This PR is exploratory and not solution discussed with the community.*

## Description

Functions can now mount volumes with this PR

## Motivation and Context

- [ ] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

I needed the functionality and decided that this can be used by someone inspiration as part of the profiles support: #586 

Also was suggested in:
https://github.com/openfaas/faas/issues/1178
and:
https://github.com/openfaas/faas/issues/320

This POC does exactly that.

## How Has This Been Tested?

Image in:
`martindekov/faas-netes/0.0.2`

With function applying the following annotations:
```
...
    annotations:
      com.openfaas.volumename: "some-pv-storage"
      com.openfaas.pvcname: "some-claim"
      com.openfaas.pvcmountpath: "/some/directory"
...
```

The function's deployment had those fields:
```
...
        volumeMounts:
        - mountPath: /some/directory
          name: some-pv-storage
        - mountPath: /var/openfaas/secrets
          name: some-projected-secrets
          readOnly: true
      dnsPolicy: ClusterFirst
...
      volumes:
      - name: some-pv-storage
        persistentVolumeClaim:
          claimName: some-claim

...
```

And the volume was mounted successfully.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] POC not to be integrated or

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
Just a thought here. If this was added, would it cover every use-case if only one volume was supported per function?

File: handlers/deploy.go [hunk 1]
```diff
@@ -142,12 +142,27 @@ func makeDeploymentSpec(request types.FunctionDeployment, existingSecrets map[st
 	annotations := buildAnnotations(request)
 
 	var serviceAccount string
+	var volumeName string
```
Comment (alexellis) [inline handlers/deploy.go]: You'd have to handle updates too, if someone wanted to remove a volume that was already attached.

File: handlers/deploy.go [hunk 2]
```diff
@@ -207,11 +222,27 @@ func makeDeploymentSpec(request types.FunctionDeployment, existingSecrets map[st
 							SecurityContext: &corev1.SecurityContext{
 								ReadOnlyRootFilesystem: &request.ReadOnlyRootFilesystem,
 							},
+							VolumeMounts: []apiv1.VolumeMount{
```
Comment (alexellis) [inline handlers/deploy.go]: This adds a volume even if one isn't configured, so you should set up the slice outside of this piece of code. ideally in a method to create it.

File: handlers/deploy.go [hunk 3]
```diff
@@ -207,11 +222,27 @@ func makeDeploymentSpec(request types.FunctionDeployment, existingSecrets map[st
...
+									Name:      volumeName,
+									MountPath: pvcMountPath,
+								},
+							},
 						},
 					},
 					ServiceAccountName: serviceAccount,
 					RestartPolicy:      corev1.RestartPolicyAlways,
 					DNSPolicy:          corev1.DNSClusterFirst,
+					Volumes: []apiv1.Volume{
```
Comment (alexellis) [inline handlers/deploy.go]: Same as comment above

**Comment by martindekov** [comment]
>Just a thought here. If this was added, would it cover every use-case if only one volume was supported per function?
Added this as POC if we want to merge it in master I can bring this in a better state.

**Comment by martindekov** [comment]
I will clear the shortcomings. I am aware of those but was not sure whether we want this to enter master. I suppose we do want to integrate the change and I will update the PR and fix all issues you mentioned. For the question whether only one volume is supported per function - yes this is the main idea. We might be able to extend this with wildcard annotation. So the key of the annotation itself suggest the volume name path etc. But I believe this would not be needed, at least I don't need more than one volume mounted.

**Comment by alexellis** [comment]
I think you should work to fix these basic issues, since you've gone ahead with a public PR, we can at least have it at a quality that anyone who lands here can test it or try it out. Would you be open to fixing up the initial feedback?

**Comment by martindekov** [comment]
I will close this, but will continue refining it and will keep the change in the branch.

**Pull request closed without merging**